### ISSUE TITLE
html validator

### ISSUE THREAD

**Issue by stevenvachon** [inikulin/parse5#55]
It goes beyond the scope of parse5 a little, but because it's a compliant parser, if it could also provide the errors in the markup, it could be used as a validator.

**Comment by inikulin** [comment]
I think this task goes far beyond the parse5 goals. But you can try build validator on top of the parse5. I can expose everything that is required.

**Comment by stevenvachon** [comment]
Definitely. If the errors encountered by parse5 were emitted/etc, that'd cover the parsing errors. Further errors such as incorrect attributes and values can be done in a validator lib.

**Comment by stevenvachon** [comment]
#43 will likely be necessary.

For the record, I have no immediate goals to write an html validator, but not completely against it when my other projects are complete.

**Comment by domenic** [comment]
I might be wrong but the great innovation of HTML5 is that there are no parsing errors; everything can be parsed.

**Comment by stevenvachon** [comment]
Parsing and validating are different. For example, common Angular use of `ng-something` is invalid, but it does work. Technically correct use of Angular directives would be `data-hg-something`.

**Comment by inikulin** [comment]
@domenic Validation is a different beast. HTML5 can swallow any input, but the output maybe quite different from that you expect.

**Comment by aredridel** [comment]
Indeed -- the spec lists a whole list of validations. It's not actually that hard to put them in -- I started to in the `html5` package. They are known points in the parsing where if you apply a fixup, you emit a warning. Just add location info and that's about it. Tracking the location info within the parser is the harder part.

**Comment by inikulin** [comment]
@aredridel Right. BTW parse5 already supports location info. The only issue is that there are no error message texts provided for the error (and sometimes it's quite hard to create one). If someone from the native speakers will help me with that it will be awesome. However, I would like to pospone this functionality for now in favor of the streaming support.

**Comment by stevenvachon** [comment]
Streaming +1

**Comment by sideshowbarker** [comment]
To put this in terms of the actual language in the HTML spec, I take it this relates to what @aredridel mentions in https://github.com/inikulin/parse5/issues/55#issuecomment-119595297 —that is, the fact that the HTML spec defines the concept of a **parse error** https://html.spec.whatwg.org/multipage/syntax.html#parse-error but the reporting of parse errors is an optional feature that an implementation of the parsing algorithm can choose to support or not, and in the case that it does do reporting of parse errors, it becomes an **error-reporting parser**.

A couple of error-reporting parsers whose error-reporting code would probably be worth taking at look at:
 
* @hsivonen’s HTML parser https://github.com/validator/htmlparser Its error reporting is used when you do **View source** in Firefox (if document has parse errors, they show up in red in that view, and if you mouse over anything in red, the corresponding message for the parse error is shown).
* The Servo HTML parser, html5ever https://github.com/servo/html5ever

**Comment by sideshowbarker** [comment]
> I might be wrong but the great innovation of HTML5 is that there are no parsing errors; everything can be parsed.

I guess what you mean is there are no fatal parsing errors. I think it’s kind of unfortunate that the spec uses the term **parse error** to define something that’s neither a fatal error nor even an error that implementations are required to do anything with. I wish it had instead been named **parse exception** or **parse warning** or something instead. I think that would align better with what it is, conceptually.

**Comment by sideshowbarker** [comment]
> The only issue is that there are no error message texts provided anywhere in the spec (and sometimes it's quite hard to create one). If someone from the native speakers will help me with that then it will be awesome.

I’d think that mostly you could just re-use the error text from @hsivonen’s parser source. See the text in:
* `err(…)` calls in https://github.com/validator/htmlparser/blob/master/src/nu/validator/htmlparser/impl/ErrorReportingTokenizer.java
* `errNoCheck(…)` calls in https://github.com/validator/htmlparser/blob/master/src/nu/validator/htmlparser/impl/TreeBuilder.java

But beyond that if/when you want more help with language of the error messages, I’m very glad to help.

**Comment by inikulin** [comment]
@sideshowbarker Wow, thanks for the helpful references.

**Comment by inikulin** [comment]
I guess implementing full-featured validation will never be in scope of parse5. The best we can is to implement parsing errors, but it's blocking on https://github.com/whatwg/html/issues/1339 because we can't properly test them at the moment with html5lib. Therefore, I'm closing it for now. Feel free to open new issue once we have error codes implemented in HTML spec if you are interested in this functionality.

**Comment by stevenvachon** [comment]
Does this mean that parsing errors are not in scope either?

**Comment by inikulin** [comment]
Nope, this means that it makes no big sense to discuss this feature until we will not have standard error codes. We can use error codes from other parsers as @sideshowbarker suggested, but we will not be able to properly test it and therefore maintain them up to date.

**Comment by stevenvachon** [comment]
Can't it have its own error codes? A wrapping library can translate them.

**Comment by inikulin** [comment]
> Can't it have its own error codes?
 
->

> We can use error codes from other parsers as @sideshowbarker suggested, but we will not be able to properly test it and therefore maintain them up to date.

>  A wrapping library can translate them

I would like to implement this feature only once =)

**Comment by stevenvachon** [comment]
Precisely why I said that we could use our own codes. No need to maintain them as they'd be unique to any standard. A wrapping library (not this one) could translate them to any standard.

**Comment by RReverser** [comment]
@stevenvachon It's not just about codes; parse errors are not statically and are added/removed as the spec evolves, and due to the absence of cross-implementations unit tests it will be non-trivial task to maintain them up-to-date.

**Comment by stevenvachon** [comment]
Aye. Well, this situation sucks.
